Add array conversions

robin-nitrokey opened the pull request:
This patch adds infallible conversions from arrays of the same size and references to arrays of the same or a smaller size.

sosthene-nitrokey commented:
I'm in the process implementing what was discussed in #4

robin-nitrokey commented:
Ah, I missed the point with the `from` method.  I’ll just drop the commit from this PR.

sosthene-nitrokey reviewed (approved):
I would also add a method for converting a `Bytes<N>` to `Bytes<M>` where `M >= N`. This can't be done through the usual conversion traits, but it can still be useful.

sosthene-nitrokey commented on src/lib.rs:
```diff
@@ -211,6 +211,59 @@ impl<const N: usize> Bytes<N> {
+    fn from(bytes: &[u8; M]) -> Self {
+        struct AssertLessThanEq<const I: usize, const J: usize>;
+        impl<const I: usize, const J: usize> AssertLessThanEq<I, J> {
+            const ASSERT: () = assert!(I <= J);
```
I think you can also add a constant message to the `assert` call.

```suggestion
            const ASSERT: () = assert!(I <= J, "Cannot convert infallibly between two arrays when the capacity of the new array is not sufficient");
```

robin-nitrokey commented:
- Removed `Bytes::from` again because otherwise it is much harder to write the examples for the `From` implementations.
- Added a message to the `assert!`.
- Added `Bytes::increase_capacity`.

sosthene-nitrokey commented on src/lib.rs:
```diff
@@ -211,6 +223,60 @@ impl<const N: usize> Bytes<N> {
 
+/// Construct a `Bytes<N>` instance from an array with `N` elements.
+///
+/// Currently, the array is copied, but a more efficient implementation could be used in the
+/// future.
```
nit: Is it really copied? I think this only depends on the compiler optimizations being smart enough to avoid moving around memory, but that would also apply if we just used `transmute` for example. This depends on the stack layout so I'm not sure it's worth mentioning.

robin-nitrokey replied:
Admittedly I did not look at the generated code but I think we saw previously that `Vec::extend_from_slice` is not optimized well so I would expect a transmute to do better.  But I can also just drop that line if you prefer.

sosthene-nitrokey replied:
Nah, let's just merge.
